### DISCUSSION TITLE
Autoscoping of constructors

### DISCUSSION THREAD

**Discussion by JaroslavTulach** [enso-org/enso#8646] [Ideas and feedback]
Do you remember how ugly it is to always prefix a constructor of a type with its type name? Always write something like:
```ruby
table.aggregate Aggregate_Column.Count
```
Enso should do better! Let's discuss how to do it!

**Comment by JaroslavTulach** [comment]
One idea is to [reuse .symbol syntax](https://github.com/enso-org/enso/issues/8645#issuecomment-1871791238). It does seem to work for
```ruby
from Standard.Base import all

type X
    A
    B x

cons x:X = x.to_text

main = cons .A
```
but it doesn't work for `cons .B 3`. Opinions, ideas?

**Reply by radeusgd** [reply]
I assume it would have to be `cons (.B 3)` to make sense.

What about the `UnresolvedConstructor` approach that I was suggesting?

The `.B` can be converted not to `UnresolvedSymbol<B>` but `UnresolvedConstructor<B>` instance (we can tell it's a constructor because it starts with uppercase). Then, in function application, whenever we apply an argument, the `UnresolvedConstructor` would store a thunk of that argument, i.e. `.B 3` would eval to `UnresolvedConstructor<B>(3)`. Then, once we hit the `: X` conversion, we can actually try to instantiate `X.B` and fowrard all of the 'pending' arguments that were stored within the `UnresolvedConstructor`.

**Reply by wdanilo** [reply]
The syntax `cons .A` should be translated to `cons (x -> x.A)`. Otherwise, it would not be compatible with all other operators, and also, using such syntax is very useful. Somewhere in the language design doc there are examples of this syntax showing how we can implement lens-like accessors. So this syntax would not work here and also should be implemented to create such lambda accessors. In fact, I made a simplification above. `.foo` should be treated as a lens that allows BOTH accessing and setting a value, for example:

```
type Animal
    Cat age

lens = .age 

test x =
    Console.log (lens x) 
    lens x = 5 # Does not have to work for now.
```

The lens-setters do not have to work right now, as they can get complex in situations when they get composed or are updating nested fields. 

Anyway, setting lenses aside, we have to have a syntax for auto-scoping. I propose the following one:

```
main = cons ...A
```

Which clearly shows that something is in front of A and we do not tell what. What do you think of it?

**Reply by wdanilo** [reply]
Another alternative would be to add a keyword for that and use it as `auto.C` - this might be even better, considering that we already have `...` syntax for postponing applying default args, like `foo = bar 1 2 ...`

**Reply by wdanilo** [reply]
Radek proposed `..C` and its even better than `...C`. I think the best bets are `..C` and `auto.C`. I like them both. I agree with Radek that `..C` is so much shorter and nicer in the code, that probably its the way to go.

**Reply by JaroslavTulach** [reply]
Looks like we really need new syntax. Trying to fit into `.A` (even if we differentiate between lowercase and upper case as @radeusgd  suggests) is probably not good idea. `...` isn't bad, but it is long. My favorite would be `~`:
```ruby
main = cons ~A
```
`~` is already used for _suspended executions_.  Using it for _suspended resolution_ might not be that bad idea.

-1 on using `auto.` - the goal is to shorten the syntax and `auto.` prefix is too long.

**Reply by wdanilo** [reply]
So, we had discussion yesterday. The `~` operator can't be used there because of a meaning clash. Enso was always designed in such a way, that type-level expressions can be processed on value-level, and vice versa. For example, you should be able to write in the future something along the lines of:

```
extract a = case a of 
    \Vector x -> x
    \List x -> x

my_fn: _a -> extract _a
my_fn = ...
```

It means, that you should be able to process types on value level, being able to pattern match and construct them. We had already problem with the `->` operator here and that's why there is another ticket to change the lambda expression from `a -> b` to `\a -> b`. For example, imagine that you want to pattern match on the provided function type:

```
extract_second_arg fn = case fn of
    \(a -> b -> c) -> b
    \_ -> Nothing
```

Please note, that the expression in parentheses `(a -> b -> c)` does not use `\`, so it is a pattern match on a function type constructor, not a lambda expression. Exactly the same applies to the proposed `~` syntax. It means laziness on type level. So Imagine we have a function:

```
fn: Vector ~_a -> Int 
fn = ...
```

And we want to extract its type to another function that will be called on type level, so we will refactor the above code to:

```
foo = Vector ~_a -> Int 

fn: foo
fn = ...
```

Now, the `foo` expression will be invalid, as `~` will mean auto-scoping, not a lazy arg constructor.

**Reply by JaroslavTulach** [reply]
- Tracking issue is #9275

**Reply by JaroslavTulach** [reply]
- #9285 has just been integrated

**Comment by JaroslavTulach** [comment]
Pull request 
- #9190 

is ready for your consideration.

**Comment by JaroslavTulach** [comment]
## Case Constructor Patterns

Will there be any change to `case` constructor patterns? For example [here](https://github.com/enso-org/enso/blob/259ad09957ca2dc262e9da4e2a644e70d0d02f42/test/Benchmarks/src/Sieve/Sieve_Original.enso#L22) one might be tempted to write:
```ruby
case f of
           ~Item p next -> if n % p == 0 then False else
                if p * p > n then True else @Tail_Call iterate next
            ~Empty -> True
```
but I guess that's not really feasible - we _need exact identification of the constructor_ before matching anything and that wouldn't be possible. All we could match would be the name and arity of the constructor and that doesn't seem to be enough.

**Reply by radeusgd** [reply]
Interesting feature, it gives me vague Lisp-like vibes.

It's pretty orthogonal to the original use-case though. In matching I think we are fine with the FQN and indeed we want to know the full type. In the future maybe we could infer this in the type system (if I know that `f : My_Type` then I could assume that `~Item` refers to `My_Type.Item`). But I don't think it adds any value for us now so I'd leave this until we find any use cases for that.

[the remark this answers is not in the document]
**Reply by JaroslavTulach** [reply]
That's a very good question! I think that we can begin using it in _dynamic drop downs_ - then the decision to use the `~` form is in the hands of the libraries team. Probably there is no need for a work on the IDE side.

[the remark this answers is not in the document]
**Reply by JaroslavTulach** [reply]
It'd be good if the `~` form worked in _static drop downs_ as well. I believe we need some co-operation with _suggestion database_ in such cases, CCing @4e6. When there is a following `Counting` type:
```ruby
type Counting
    Zero
    One
    Two 
    Many n:Integer
```
then a method with **signature defined inline** like
```ruby
next c:Counting = case c of
   Counting.Zero -> Counting.One
   Counting.Two -> Counting.Many 3
   Counting.Many n -> Counting.Many n+1
``` 
is guaranteed to accept short form of `~Zero`, `~One` or `~Many 5`. The _suggestion database_ shall indicate this to the IDE.

On the other hand a method with _no signature_ or _signature defined on a separate line_ (see https://github.com/enso-org/enso/issues/6884#issuecomment-1938119826 for details) like
```ruby
prev : Counting -> Counting 
prev c = ...
```
**doesn't guarantee** to accept the short form. Again _suggestion database_ shall provide such information to the IDE. What do you think, Dmitry?

[the remark this answers is not in the document]
**Reply by 4e6** [reply]
Could you explain in more details how the suggestions database is involved here?

[the remark this answers is not in the document]
**Reply by JaroslavTulach** [reply]
It'd be good if the `~` form worked in _static drop downs_. I believe we need some co-operation with _suggestion database_ to provide the IDE information whether it is possible to use _autoscoped constructors_ or not for a given function argument.

[the remark this answers is not in the document]
**Reply by JaroslavTulach** [reply]
Dmitry reported 
- #9259

as a followup issue.

**Comment by kazcw** [comment]
How does this relate to the current `Tree.AutoScope` syntax, e.g. here:

https://github.com/enso-org/enso/blob/e98306f170c0f435f4783a2bdb42801af2b48706/distribution/lib/Standard/Base/0.0.0-dev/src/Meta.enso#L138-L140

**Reply by radeusgd** [reply]
The syntax is wrongly named, `...` is used for suspending default arguments, so it should be called something else, e.g. `Tree.SuspendedDefaultArguments`.
